src/forms/renderers/OAuth/index.tsx

```diff
@@ -95,146 +78,50 @@ const OAuthproviderRenderer = ({
-                return true;
-            }
-
-            // When inside a oneOf/anOf/etc. we need to check we have the discriminator
-            if (discriminatorProperty && field === discriminatorProperty) {
-                return true;
-            }
-
-            // Check if the field is in the main data
-            if (includes(dataKeys, field) && hasLength(data?.[field])) {
-                return true;
-            }
-
-            // As a last effort check if the field is inside of nested data
-            if (
-                includes(nestedKeys, field) &&
-                hasLength(data?.[onChangePath]?.[field])
-            ) {
-                return true;
-            }
-
-            // Didn't find anything - return false
-            return false;
-        });
-
-        setHasAllRequiredProps(response);
-    }, [data, discriminatorProperty, onChangePath, requiredFields]);
-
-    // Pull out the provider so we can render the button
-    const providerVal = options ? options[Options.oauthProvider] : NO_PROVIDER;
-    const provider = useMemo(() => startCase(providerVal), [providerVal]);
-
-    // handler for the useOauth stuff
-    const onError = (error: any) => {
-        if (error === 'access_denied') {
-            setErrorMessage(
-                intl.formatMessage(
-                    {
-                        id: 'oauth.authentication.denied',
-                    },
-                    { provider }
-                )
-            );
-        } else {
-            setErrorMessage(error);
-        }
-    };
-
-    // handler for the useOauth stuff
-    const onSuccess = async (payload: any) => {
-        const tokenResponse = await accessToken(
-            payload.state,
-            payload.code,
-            endpointConfigData
-        );
+    const { hasAllRequiredProps, showAuthenticated } = useAllRequiredPropCheck(
+        data,
+        options,
+        onChangePath,
+        discriminatorProperty
+    );
 
-        if (tokenResponse.error) {
-            setErrorMessage(
-                intl.formatMessage(
-                    { id: 'oauth.accessToken.error' },
-                    { provider }
-                )
-            );
-        } else if (isEmpty(tokenResponse.data)) {
-            setErrorMessage(
-                intl.formatMessage(
-                    { id: 'oauth.emptyData.error' },
-                    { provider }
-                )
-            );
-        } else {
-            handleChange(onChangePath, {
-                ...fakeDefaults,
+    // OAuth Stuff
+    const { errorMessage, openPopUp, loading } = useOauthHandler(
+        provider,
+        (tokenResponse) => {
+            // Order matters here
+            //  1. Get the data we have so far
+            //  2. Merge in the injected values as those MUST match what the server is expecting
+            //  3. Add in the response from the server for the access token
+            const updatedCredentials = {
```

Comment:
Moved the logic over to a hook to keep things a bit cleaned inside this huge file.

Reply:
The success handler here merges in the injected values again so that they are there when truly needed. This fixes the issue from https://github.com/estuary/ui/issues/513